### PR TITLE
Performance optimization 1

### PR DESCRIPTION
**Pull request by halil81**
This PR converts some Iterables to List and also upgrades some method signatures.

### CODE REVIEW

File: lib/src/serializers/json_api.dart [hunk 1]
```diff
@@ -19,12 +19,15 @@ class JsonApiSerializer implements Serializer {
   @override
   JsonApiManyDocument deserializeMany(String payload) {
     Map<String, dynamic> parsed = parse(payload);
-    var docs = (parsed['data'] as Iterable).map((item) => JsonApiDocument(
-        item['id'],
-        item['type'],
-        item['attributes'],
-        item['relationships'],
-        parsed['included']));
+    var docs = (parsed['data'] as List)
```
Comment (algodave) [inline lib/src/serializers/json_api.dart:22]: @halil81 Please reinstate `as Iterable` in this case. We're navigating a structure created by `dart:convert` so we can't make assumptions on the exact type.

File: lib/src/serializers/json_api.dart [hunk 2]
```diff
@@ -19,12 +19,15 @@ class JsonApiSerializer implements Serializer {
...
+        .map((item) => JsonApiDocument(
+              item['id'],
+              item['type'],
+              item['attributes'],
+              item['relationships'],
+              parsed['included'],
+            ))
+        .toList();
     return JsonApiManyDocument(docs, parsed['included'], parsed['meta']);
```
Comment (algodave) [inline lib/src/serializers/json_api.dart:31]: @halil81 We can't be sure that `parsed['included']` will return a `List`, so let's use `List.from()`

**Review by algodave — CHANGES_REQUESTED** [review]

File: lib/src/serializers/json_api.dart (hunk 1 above)
Comment (halil81) [reply lib/src/serializers/json_api.dart]: Thanks! @algodave, docs reinstated as `Iterable`

File: lib/src/serializers/json_api.dart (hunk 2 above)
Comment (halil81) [reply lib/src/serializers/json_api.dart]: Thanks! @algodave, `parsed['included']` wrapped in `List.from()`

**Comment by algodave** [comment]
@halil81 Let's update the signature of the following methods that are implemented in `JsonApiDocument`
 and proxied in `JsonApiModel`:

 - `idsFor`
 - `includedIdsFor` 
 - `errorsFor`

so that the return type is a `List` instead of an `Iterable`.

**Comment by algodave** [comment]
@halil81 I still see a lot of `Iterable` usages:

Screen Shot 2022-01-26 at 11 18 32 AM

Most of those should be very easy to replace with `List`, can we convert the easy ones in this PR? 

It would also be great to list the "hard" ones in a comment here, so we can discuss a conversion strategy. Thanks!

**Comment by halil81** [comment]
@algodave I've converted most of the remaining `Iterable` and the ones that are hard to convert for now are below. Thanks!

In `Adapter` class:

- findMany()
- findAll()
- query()
- peekMany()
- peekAll()

**Comment by algodave** [comment]
Thanks @halil81 well done! I agree that we can leave conversion of other `Iterable`s to a later iteration, because they all depend on the following design decision:

`JsonApiManyDocument extends Iterable<JsonApiDocument>`

With the changes in this PR, `JsonApiManyDocument` now holds a `List<JsonApiDocument>`. That should address performance issues, so we could even decide to leave it extending `Iterable`. We'll test and see 🙂

**Pull request merged**